// slate-trailing-block/dist/index.js
'use strict';

var Slate = require('slate');
var focusAtEnd = require('./focusAtEnd');

/**
 * Slate plugin to ensure a trailing block.
 * @param {Object} [opts] Options for the plugin
 * @param {String|Function} [opts.match='paragraph'] Match last block
 * @param {String} [opts.type] The type of the trailing block to insert
 * @return {Object}
 */

function TrailingBlock(opts) {
    opts = opts || {};
    opts.type = opts.type || 'paragraph';
    opts.match = opts.match || function (node) {
        return node.type === opts.type;
    };

    return {
        schema: {
            rules: [{
                match: function match(node) {
                    return node.kind === 'document';
                },
                validate: function validate(node) {
                    var lastNode = node.nodes.last();

                    return !lastNode || !opts.match(lastNode) ? true : null;
                },
                normalize: function normalize(transform, node, value) {
                    var lastIndex = node.nodes.count();
                    var block = Slate.Block.create({
                        type: opts.type,
                        nodes: [Slate.Text.create()]
                    });

                    return transform.insertNodeByKey(node.key, lastIndex, block);
                }
            }]
        },

        transforms: {
            focusAtEnd: focusAtEnd
        }
    };
}

module.exports = TrailingBlock;
