// slate-drop-or-paste-images/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _es6Promise = require('es6-promise');

var _es6Promise2 = _interopRequireDefault(_es6Promise);

var _isImage = require('is-image');

var _isImage2 = _interopRequireDefault(_isImage);

var _isUrl = require('is-url');

var _isUrl2 = _interopRequireDefault(_isUrl);

var _mimeTypes = require('mime-types');

var _mimeTypes2 = _interopRequireDefault(_mimeTypes);

var _path = require('path');

var _loadImageFile = require('./load-image-file');

var _loadImageFile2 = _interopRequireDefault(_loadImageFile);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Insert images on drop or paste.
 *
 * @param {Object} options
 *   @property {Function} applyTransform
 *   @property {Array} extensions (optional)
 * @return {Object} plugin
 */

function DropOrPasteImages(_ref) {
  var applyTransform = _ref.applyTransform,
      extensions = _ref.extensions;

  if (!applyTransform) {
    throw new Error('You must supply an `applyTransform` function.');
  }

  /**
   * Apply the transform for a given file and update the editor with the result.
   *
   * @param {Transform} transform
   * @param {Editor} editor
   * @param {Blob} file
   * @return {Promise}
   */

  function asyncApplyTransform(transform, editor, file) {
    return _es6Promise2.default.resolve(applyTransform(transform, file)).then(function () {
      var next = transform.apply();
      editor.onChange(next);
    });
  }

  /**
   * On drop or paste.
   *
   * @param {Event} e
   * @param {Object} data
   * @param {State} state
   * @param {Editor} editor
   * @return {State}
   */

  function onInsert(e, data, state, editor) {
    switch (data.type) {
      case 'files':
        return onInsertFiles(e, data, state, editor);
      case 'html':
        return onInsertHtml(e, data, state, editor);
      case 'text':
        return onInsertText(e, data, state, editor);
    }
  }

  /**
   * On drop or paste files.
   *
   * @param {Event} e
   * @param {Object} data
   * @param {State} state
   * @param {Editor} editor
   * @return {State}
   */

  function onInsertFiles(e, data, state, editor) {
    var target = data.target,
        files = data.files;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {

      for (var _iterator = files[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var file = _step.value;

        if (extensions) {
          var ext = _mimeTypes2.default.extension(file.type);
          if (!extensions.includes(ext)) continue;
        }

        var transform = state.transform();
        if (target) transform.moveTo(target);

        asyncApplyTransform(transform, editor, file);
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return state;
  }

  /**
   * On drop or paste html.
   *
   * @param {Event} e
   * @param {Object} data
   * @param {State} state
   * @param {Editor} editor
   * @return {State}
   */

  function onInsertHtml(e, data, state, editor) {
    var html = data.html,
        target = data.target;

    var parser = new DOMParser();
    var doc = parser.parseFromString(html, 'text/html');
    var body = doc.body;
    var firstChild = body.firstChild;
    if (firstChild.nodeName.toLowerCase() != 'img') return;

    var src = firstChild.src;

    if (extensions) {
      var ext = (0, _path.extname)(src).slice(1);
      if (!extensions.includes(ext)) return;
    }

    (0, _loadImageFile2.default)(src, function (err, file) {
      if (err) return;
      var transform = editor.getState().transform();
      if (target) transform.moveTo(target);
      asyncApplyTransform(transform, editor, file);
    });

    return state;
  }

  /**
   * On drop or paste text.
   *
   * @param {Event} e
   * @param {Object} data
   * @param {State} state
   * @param {Editor} editor
   * @return {State}
   */

  function onInsertText(e, data, state, editor) {
    var text = data.text,
        target = data.target;

    if (!(0, _isUrl2.default)(text)) return;
    if (!(0, _isImage2.default)(text)) return;

    (0, _loadImageFile2.default)(text, function (err, file) {
      if (err) return;
      var transform = editor.getState().transform();
      if (target) transform.moveTo(target);
      asyncApplyTransform(transform, editor, file);
    });

    return state;
  }

  /**
   * Return the plugin.
   *
   * @type {Object}
   */

  return {
    onDrop: onInsert,
    onPaste: onInsert
  };
}

/**
 * Export.
 *
 * @type {Function}
 */

exports.default = DropOrPasteImages;
